File: 0057-insert-interval/0057-insert-interval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
const overlap = (x, y) => !(x[1] < y[0] || y[1] < x[0]);

const insert = (intervals, newInterval) => {
    const n = intervals.length;
    let res = [];
    let i = 0;
    while (i < n && intervals[i][1] < newInterval[0]) {
        res.push(intervals[i]);
        i++;
    }
    
    while (i < n && overlap(intervals[i], newInterval)) {
        newInterval = [Math.min(intervals[i][0], newInterval[0]), Math.max(intervals[i][1], newInterval[1])];
        i++;
    }
    
    res.push(newInterval);
    
    while (i < n) {
        res.push(intervals[i]);
        i++;
    }
    
    return res;
};

File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        intervals.append(newInterval)
        def sortBy(interval):
            return interval[0]
        intervals.sort(key=sortBy)
        def overlap(a, b):
            return not (a[1] < b[0] or b[1] < a[0])
        def merge(intervals):
            merged = [intervals[0]]
            i = 1
            while i < len(intervals):
                if overlap(merged[-1], intervals[i]):
                    merged[-1] = [min(merged[-1][0], intervals[i][0]), max(merged[-1][1], intervals[i][1])]
                else:
                    merged.append(intervals[i])
                i += 1
            return merged
        return merge(intervals)